### PR TITLE
8235385: Crash on aarch64 JDK due to long offset

### PR DESCRIPTION
**Pull request by apavlyutkin**
Hi!

Please review "backport" of [Wang Zhuo's patch](http://cr.openjdk.java.net/~wzhuo/BigOffsetAarch64/webrev.00/jdk13u.patch) to 11u-dev. The patch is applied with the following adaptation:

- the legitimizing overload
```
static void loadStore(MacroAssembler masm, mem_insn insn,
                         Register reg, int opcode,
                         Register base, int index, int size, int disp,
                         int size_in_memory);
```
ported from 15u-dev with slight changes
-  `aarch64_enc_strh`, `aarch64_enc_strh0`, `aarch64_enc_strw`, `aarch64_enc_strw0`, `aarch64_enc_str`, and `aarch64_enc_str0` functions updated to use new loadStore() overload

Verified (18.04.6 LTS/aarch64) with the reproducers from JBS, 10 of 10 runs passed. Aproximately a half of runs crashed before I've applied the patch.

Regression (18.04.6 LTS/aarch64): hotspot_compiler

@theRealAph @AntonKozlov please take a look at that. Andrew, I could not mention Wang, if you know his ID could you mention him as well. Thank you

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8235385](https://bugs.openjdk.java.net/browse/JDK-8235385): Crash on aarch64 JDK due to long offset

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Contributors
 * Wang Zhuo `<wzhuo@openjdk.org>`
 * Andrew Haley `<aph@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk11u-dev pull/960/head:pull/960` \
`$ git checkout pull/960`

Update a local copy of the PR: \
`$ git checkout pull/960` \
`$ git pull https://git.openjdk.java.net/jdk11u-dev pull/960/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 960`

View PR using the GUI difftool: \
`$ git pr show -t 960`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk11u-dev/pull/960.diff">https://git.openjdk.java.net/jdk11u-dev/pull/960.diff</a>

</details>

### CODE REVIEW

**Comment by theRealAph** [comment]
That looks perfect, but I'd take out the `assert` because in this old release an out-of-range access isn't unexpected, and this is a good-enough way to fix it.

Co-author is @sandlerwang .

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 1]
```diff
@@ -2622,25 +2656,25 @@ encode %{
...
     loadStore(MacroAssembler(&cbuf), &MacroAssembler::strh, src_reg, $mem->opcode(),
-               as_Register($mem$$base), $mem$$index, $mem$$scale, $mem$$disp);
+               as_Register($mem$$base), $mem$$index, $mem$$scale, $mem$$disp, 2);
   %}
 
   enc_class aarch64_enc_strh0(memory mem) %{
     MacroAssembler _masm(&cbuf);
     loadStore(_masm, &MacroAssembler::strh, zr, $mem->opcode(),
-               as_Register($mem$$base), $mem$$index, $mem$$scale, $mem$$disp);
+               as_Register($mem$$base), $mem$$index, $mem$$scale, $mem$$disp, 2);
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64.ad:2665]: This stuff is all generated from `ad_encode.m4`. Please edit that file, then regenerate everything between

`  // BEGIN Non-volatile memory access`

and

`  // END Non-volatile memory access`

**Review by theRealAph — APPROVED** [review]
Thanks, that looks good. I think it's the best way to fix the bug and it doesn't touch anything unnecessarily.

**Comment by apavlyutkin** [comment]
@theRealAph Please do not sponsor the PR. We run strict testing on it and observe few crashes. We need a little time to check if they relate to the fix

**Comment by theRealAph** [comment]
> @theRealAph Please do not sponsor the PR. We run strict testing on it and observe few crashes. We need a little time to check if they relate to the fix

OK

**Comment by apavlyutkin** [comment]
> > @theRealAph Please do not sponsor the PR. We run strict testing on it and observe few crashes. We need a little time to check if they relate to the fix
> 
> OK

Double-checked: the same our tests fails on master, thus the crashes do not relate to the fix. Please sponsor the PR

**Pull request closed without merging**